crates/red_knot_python_semantic/src/types.rs

```diff
@@ -29,6 +29,86 @@ mod display;
 mod infer;
 mod narrow;
 
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum Boundedness {
+    Bound,
+    MayBeUnbound,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum SymbolLookupResult<'db> {
+    Type(Type<'db>, Boundedness),
```

Comment:
The flat representation also seemed more obvious to me at first too -- my first instinct would be to do something like this:

```rs
enum SymbolLookupResult<'db> {
    /// The symbol is not bound in any code paths.
    ///
    /// All attempts to read the symbol result in a diagnostic.
    /// From the perspective of other expressions and statements,
    /// the symbol has type `Unknown`.
    AlwaysUnbound,

    /// The symbol is bound in some code paths, but not in others;
    /// `bound_ty` gives the type of the symbol in the paths in which it is bound.
    ///
    /// All attempts to read the symbol result in a diagnostic,
    /// but one that conveys less certainty than for symbols
    /// which are always unbound. From the perspective of other expressions
    /// and statements, the symbol has type `bound_ty`.
    PossiblyBound { bound_ty: Type<'db> },

    /// The symbol is bound in all code paths;
    /// reading the symbol never results in a diagnostic.
    AlwaysBound { ty: Type<'db> },
}
```

But I can see that the nested representation is much more extensible for the future, when we will also want to incorporate declaredness into `SymbolLookupResult`.

Throwing yet another option into the mix, did you consider something like this? The disadvantage of it is that there is an "implicit invariant" that `bound_ty` would always be `None` if `boundedness` was `Boundedness::Unbound`. But it would be highly extensible, since we could simply add `declared_ty` and `declaredness` fields later:

```rs
enum Boundedness {
    /// The symbol is unbound in all code paths
    AlwaysUnbound,
    /// The symbol is bound in some code paths, unbound in others
    PossiblyBound,
    /// The symbol is bound in all code paths
    AlwaysBound
}

struct SymbolLookupResult<'db> {
    /// The bound type of the symbol.
    ///
    /// If `boundedness is `AlwaysUnbound`, this will be `None`;
    /// else, this will be equivalent to the union of the types
    /// in all code paths where the symbol is bound
    bound_ty: Option<Type<'db>>,

    boundedness: Boundedness,
}
```

Reply:
> Throwing yet another option into the mix, did you consider something like this?

To be honest: no :smile:. Give up the type-level safety for … what? Future extensibility... ? It's a bit hard for me to tell what we need here in the future, because I don't know how that declared-ness information will come in eventually. I built a simple solution for what we have now, but I'm open for change proposals, if you think it's something that needs to be addressed here.